src/net/server/stream.rs

```diff
@@ -0,0 +1,667 @@
+use std::string::{String, ToString};
+use std::sync::{Arc, Mutex};
+use tokio::net::TcpListener;
+use tokio::sync::watch;
+use tokio::time::{interval, timeout, MissedTickBehavior};
+use tracing::{error, trace, trace_span};
+
+use crate::net::server::buf::BufSource;
+use crate::net::server::error::Error;
+use crate::net::server::metrics::ServerMetrics;
+use crate::net::server::middleware::chain::MiddlewareChain;
+use crate::net::server::service::Service;
+use crate::net::server::sock::AsyncAccept;
+use crate::utils::config::DefMinMax;
+
+use super::buf::VecBufSource;
+use super::connection::{self, Connection};
+use super::service::ServerCommand;
+use crate::base::wire::Composer;
+
+// TODO: Should this crate also provide a TLS listener implementation?
+
+/// A TCP transport based DNS server.
+///
+/// The TCP aka Transport Control Protocol, as [noted by Wikipedia], is a
+/// stream based transport protocol. This type defines a type of
+/// [`StreamServer`] that expects connections to be received via
+/// [`tokio::net::TcpListener`] and can thus be used to implement a TCP based
+/// DNS server.
+///
+/// [noted by Wikipedia]:
+///     https://en.wikipedia.org/wiki/Reliable_byte_streamuse
+/// [`tokio::net::TcpListener`]:
+///     https://docs.rs/tokio/latest/tokio/net/struct.TcpListener.html
+pub type TcpServer<Svc> = StreamServer<TcpListener, VecBufSource, Svc>;
+
+/// Limit on the number of concurrent TCP connections that can be handled by
+/// the server.
+///
+/// The value has to be between one and 100,000. The default value is 100. The
+/// default value is based on the default value of the NSD 4.8.0 `-n number`
+/// configuration setting .
+///
+/// If the limit is hit, further connections will be accepted but closed
+/// immediately.
+const MAX_CONCURRENT_TCP_CONNECTIONS: DefMinMax<usize> =
+    DefMinMax::new(100, 1, 100000);
+
+//----------- Config ---------------------------------------------------------
+
+/// Configuration for a stream server connection.
+pub struct Config<RequestOctets, Target>
+where
+    RequestOctets: Octets,
+    Target: Composer + Default,
+{
+    /// Limit on the number of concurrent TCP connections that can be handled
+    /// by the server.
+    pub(super) max_concurrent_connections: usize,
+    pub(super) connection_config: connection::Config<RequestOctets, Target>,
+}
+
+impl<RequestOctets, Target> Config<RequestOctets, Target>
+where
+    RequestOctets: Octets,
+    Target: Composer + Default,
+{
+    /// Creates a new, default config.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Set the limit on the number of concurrent TCP connections that can be
+    /// handled by the server.
+    ///
+    /// The value has to be between one and 100,000. The default value is 100.
+    /// The default value is based on the default value of the NSD 4.8.0 `-n
+    /// number` configuration setting .
+    ///
+    /// If the limit is hit, further connections will be accepted but closed
```

Comment:
Why accept connections and then close them?

Reply:
As the actual accepting of connections is handled by the `AsyncAccept` trait impl and not by the server itself the server has no way to avoid accepting connections. It can however close them if it decides it has too many accepted connections.

Note also that Unbound has a [`tcp-connection-limit`](https://unbound.docs.nlnetlabs.nl/en/latest/manpages/unbound.conf.html#unbound-conf-tcp-connection-limit) configuration setting which is described as _"When at the limit, further connections are accepted but closed immediately"_.

Reply:
Done.